src/test/rgw/test_d4n_filter.cc

```diff
@@ -0,0 +1,2819 @@
+      request req;
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", TEST_BUCKET + testName + "__:null_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_9");
+      req.push("HGET", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_0", "version"); 
+
+      response< int, int, int, std::string > resp;
+
+      conn->async_exec(req, resp, yield[ec]);
+      ASSERT_EQ((bool)ec, false);
+      EXPECT_EQ((int)std::get<0>(resp).value(), 1);
+      EXPECT_EQ((int)std::get<1>(resp).value(), 1);
+      EXPECT_EQ((int)std::get<2>(resp).value(), 1);
+      
+      version = std::get<3>(resp).value();
+    }
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + version, err), true);  
+    std::string oid = version + "#0#" + std::to_string(ofs);
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + oid, err), true);     
+
+    std::unique_ptr<rgw::sal::Object::DeleteOp> del_op = obj->get_delete_op();
+    EXPECT_EQ(del_op->delete_obj(env->dpp, optional_yield{yield}, rgw::sal::FLAG_LOG_OP), 0);
+
+    {
+      boost::system::error_code ec;
+      request req;
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", TEST_BUCKET + testName + "__:null_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_9");
+      req.push("FLUSHALL");
+
+      response< int, int, int, boost::redis::ignore_t > resp;
+
+      conn->async_exec(req, resp, yield[ec]);
+      ASSERT_EQ((bool)ec, false);
+      EXPECT_EQ((int)std::get<0>(resp).value(), 0);
+      EXPECT_EQ((int)std::get<1>(resp).value(), 0);
+      EXPECT_EQ((int)std::get<2>(resp).value(), 0);
+    }
+
+    /* TODO: Eviction cycle to delete cache blocks
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + version, err), false);  
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + oid, err), false);     
+    */
+
+    conn->cancel();
+    testBucket->remove(env->dpp, true, optional_yield{yield});
+    DriverDestructor driver_destructor(static_cast<rgw::sal::D4NFilterDriver*>(driver));
+  }, rethrow);
+
+  io.run();
+}
+
+// Read cache tests, versioned
+TEST_F(D4NFilterFixture, PutVersionedObjectRead)
+{
+  const std::string testName = "PutVersionedObjectRead";
+ 
+  boost::asio::spawn(io, [this, &testName] (boost::asio::yield_context yield) {
+    init_driver(yield);
+    create_bucket(testName, yield);
+    std::string instance;
+    put_version_enabled_object(testName, instance, yield);
+    put_version_suspended_object(testName, yield);
+
+    // TODO: No unversioned head object set in redis, is this intended?
```

Comment:
TODO: Do unversioned head objects get set in the directory for versioned objects?

Reply:
Yes, now we fetch latest version from backend store for versioned clean objects due to complexities in maintaining versions between the cache and backend store.

Reply:
Is "yes" in response to the GitHub comment or the code comment? They are phrased differently so I wanted to confirm

Reply:
I think both mean the same, don't they? If not please clarify. I meant that in case of clean objects belonging to versioned buckets do not have a latest and null hash entry in redis.